Sidebar: Add Site Checklist Progress to My Home

Aurorum opened the pull request:
#### Changes proposed in this Pull Request

* Adds a progress bar to the sidebar to convey how much progress has been made in the Checklist

#### Testing instructions

Verify the progress bar conveys the number of tasks completed, and that it looks okay on all colour schemes. It should also be hidden once the checklist is complete.

<img width="1658" alt="Screenshot 2020-05-10 at 11 02 19" src="https://user-images.githubusercontent.com/43215253/81496233-2b801300-92ae-11ea-82ed-5177985b66ce.png">

cc @mmtr, @gwwar 

Fixes #41929

Aurorum committed:
Styling for checklist progress

Aurorum committed:
Include checklist

Aurorum commented on client/my-sites/sidebar/index.jsx:
```diff
@@ -183,7 +194,14 @@ export class MySitesSidebar extends Component {
+				{ ! hideTaskProgress && (
+					<div className="sidebar__checklist-progress">
+						<p className="sidebar__checklist-progress-text">{ translate( 'Finish setup' ) }</p>
+						<ProgressBar value={ 2 } total={ taskCount } />
```
Trying to figure out what `value` should be here by calculating the number of completed tasks is proving to be the problem! I'm going to try to figure it out again, but haven't had much luck so far.

Aurorum committed:
Fix the colour on darker schemes

a8ci18n commented on client/my-sites/sidebar/index.jsx:
```diff
@@ -183,7 +194,14 @@ export class MySitesSidebar extends Component {
+			>
+				{ ! hideTaskProgress && (
+					<div className="sidebar__checklist-progress">
+						<p className="sidebar__checklist-progress-text">{ translate( 'Finish setup' ) }</p>
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **26** times:
`translate( 'Finish Setup' )` *ES Score: 12*

sixhours replied:
We'd want to stick to sentence case here, ignore the bot.

Aurorum committed:
Get progress bar working

sixhours commented:
Thanks for jumping on this, Aurorum!

I think we could safely bump the font size up to 12px from 10px for better readability, and to match the other smaller fonts to the right side of the sidebar:

<img width="679" alt="Screen Shot 2020-05-11 at 10 28 08 AM" src="https://user-images.githubusercontent.com/2124984/81573214-2c7a7880-9372-11ea-928f-e18329b7c3eb.png">

My only concern with this on the whole is there's a lot of stuff happening in that area of the sidebar; we have the upsell/nudge, the checklist nudge, the stats bar, and the plan name. Things feel pretty crowded when all of those things are active.

The nudge is also showing on a WP for Teams site that has no checklist:

<img width="1677" alt="Screen Shot 2020-05-11 at 10 34 54 AM" src="https://user-images.githubusercontent.com/2124984/81573955-2d5fda00-9373-11ea-8f03-4ab61727a2a3.png">

(You can test WP for Teams by creating a site from `/start/wp-for-teams`.)

I double-checked on all color schemes, and there are a few instances where the colors are muddy or blend in:

Sunset theme is particularly rough:
<img width="297" alt="Screen Shot 2020-05-11 at 10 24 26 AM" src="https://user-images.githubusercontent.com/2124984/81573490-91ce6980-9372-11ea-8b40-a8ced49dea3e.png">
<img width="290" alt="Screen Shot 2020-05-11 at 10 24 16 AM" src="https://user-images.githubusercontent.com/2124984/81573491-91ce6980-9372-11ea-94f2-d155b8c90183.png">

Contrast theme:
<img width="297" alt="Screen Shot 2020-05-11 at 10 25 03 AM" src="https://user-images.githubusercontent.com/2124984/81573488-9135d300-9372-11ea-90f7-1c3236f8c2fe.png">

I'm less concerned about the color schemes, though, since I realize it's going to be hard to use one color that works well with all palettes.

Aurorum committed:
Check for WPForTeams site

Aurorum committed:
Add test

Aurorum committed:
Check if site is eligible for checklist in selector

Aurorum committed:
Fix colours of progress bar

Aurorum commented:
Fixed the font size and colours (they're now simply the same as the text colour - probably the safest way to do this). 👍 

> The nudge is also showing on a WP for Teams site that has no checklist

It feels a bit odd to add the logic for that here, since it'll probably be repeated again in the future, so I've added it to the selectors. I've verified that this doesn't impact the checklist on My Home, but let me know if there's an issue with doing this!

> My only concern with this on the whole is there's a lot of stuff happening in that area of the sidebar; we have the upsell/nudge, the checklist nudge, the stats bar, and the plan name. Things feel pretty crowded when all of those things are active.

What about hiding the stats sparkline until the user has completed the checklist? The user probably won't have any views if they're still completing the checklist, so they won't be getting any useful information from it anyway.

Aurorum committed:
Fix linter (?)

kwight commented:
Hm, I thought the last version I saw had the site checklist completion progress as a linear bar above the list of items, but maybe I've missed something along the way; let's wait for @danhauk to chime in.

<img width="1255" alt="Screen Shot 2020-05-11 at 10 34 03 AM" src="https://user-images.githubusercontent.com/349751/81592734-436d9a80-9373-11ea-8f7c-4bf97d5cb681.png">

kwight commented:
> but maybe I've missed something along the way

Indeed I did, I hadn't realized the issue linked to was something new. Carry on 🙃

Aurorum commented:
It is different, yep, but I don't see the bar there either? Is this something that is worth filing a separate PR for or am I missing something too?

<img width="1019" alt="Screenshot 2020-05-11 at 19 15 53" src="https://user-images.githubusercontent.com/43215253/81596372-dd593580-93bb-11ea-9679-949d9da545db.png">

Aurorum committed:
Fix bug where Customer Home is broken for Atomic (oops)

gwwar commented:
Oh for context, the green progress indicator on the Site Setup List was something we happened to drop to help speed up the change. I think we can avoid this change on site setup list, if we see the progress indicator on the sidebar. Even if we want both it'd be fair to split that out to its own PR. cc @danhauk  

For context, the indicator on the sidebar is to help customers know how to get back to My Home after completing a task or navigating away.

Aurorum committed:
New non-risky fix for WPForTeams issue

Aurorum commented on client/my-sites/sidebar/index.jsx:
```diff
@@ -45,8 +49,11 @@ import {
+import getSiteTaskList from 'state/selectors/get-site-task-list';
 import canCurrentUserUseCustomerHome from 'state/sites/selectors/can-current-user-use-customer-home';
 import canCurrentUserManagePlugins from 'state/selectors/can-current-user-manage-plugins';
+import { getCompletionStatus } from 'lib/checklist';
```
I'm stuck on why this is causing a Lint error - it is used and it is necessary. Any ideas?

gwwar replied:
It is unused, the usage in that file is `const { completed, total } = siteTasklist.getCompletionStatus();`

gwwar replied:
I think we can simply drop this line, unless @mmtr you had suggestions on how to better fetch tasks.

Note that we may want future follow ups for modifying APIs if we'll need this information on most page loads.

mmtr replied:
> I think we can drop this line, unless @mmtr you had suggestions on how to better fetch tasks.

Yeah, this is the better way of handling it at the moment so we need to remove this import since it's unused.

Aurorum replied:
Ah, oops, you're right! Fixed. :)

gwwar reviewed (approved):
Thanks @Aurorum 💖Giving a ✅ for the code side of things. It's good to ship once lint and tests are green.

@davemart-in or @sixhours any other design requests before we land this?

This behaves pretty well, I tested on a new site, completing site-setup, checked a corner case with an older site (that had fewer checklist items), and wp teams.

<img width="1465" alt="in-progress" src="https://user-images.githubusercontent.com/1270189/81619200-dcb1a680-939d-11ea-97d2-91a0af08478e.png">
<img width="1465" alt="finished" src="https://user-images.githubusercontent.com/1270189/81619202-df140080-939d-11ea-8391-8531f0f9f594.png">
<img width="1777" alt="Screen Shot 2020-05-11 at 3 39 23 PM" src="https://user-images.githubusercontent.com/1270189/81619250-036fdd00-939e-11ea-9192-0506a5ca7ea7.png">

Aurorum committed:
Fix linter

mmtr reviewed (dismissed):
The text is a bit off when using other languages:

<img width="369" alt="Screen Shot 2020-05-12 at 13 12 01" src="https://user-images.githubusercontent.com/1233880/81679881-d7fafa00-9452-11ea-9263-a4edba5a9547.png">

<img width="282" alt="Screen Shot 2020-05-12 at 13 12 29" src="https://user-images.githubusercontent.com/1233880/81679905-dc271780-9452-11ea-8d94-f9e899091aee.png">

<img width="276" alt="Screen Shot 2020-05-12 at 13 31 25" src="https://user-images.githubusercontent.com/1233880/81682125-f104aa80-9454-11ea-8862-7f2b74967b1a.png">


But I have no idea how can we develop something that fits well in all languages. @Automattic/dotcom-manage-design do you have any suggestion?

mmtr commented:
Maybe we can wrap both the menu title and the checklist text to next line if they don't fit in a single line, and use an ellipsis if still single words don't fit in them.

```
.sidebar__menu-link-text {
  max-width: 70%;
  white-space: normal;
  padding-right: 8px;
}

.sidebar__checklist-progress-text {
  white-space: normal;
  text-overflow: ellipsis;
  overflow: hidden;
}
```

<img width="285" alt="Screen Shot 2020-05-12 at 13 42 52" src="https://user-images.githubusercontent.com/1233880/81683517-7f2d6080-9456-11ea-9f7b-9a443889700e.png">

Not the most beautiful UI but at least most of the text is visible.

davemart-in commented:
Very cool! Excellent work @Aurorum. Three minor bits of feedback from me:

- I'm maybe suggest changing the label from "Finish Setup" to be just "Setup". I think the "Finish" part can be implied. That should help with translations.
- With a shorter label, my guess is that we could reduce the width of the progress bar as well. Maybe reduce the size by 30%?
- The light gray within the progress bar against the light blue of the background clashes a bit visually. I'd suggest making that light gray portion of the progress bar much more light.

Here's a comparison:

![Image 2020-05-12 at 10 01 57 AM](https://user-images.githubusercontent.com/5634774/81701199-afb1d200-9437-11ea-9edc-9dd2a3e6058b.png)

Aurorum commented:
Thanks all, will commit the changes soon! Quick question though @davemart-in:

> The light gray within the progress bar against the light blue of the background clashes a bit visually. I'd suggest making that light gray portion of the progress bar much more light.

I've been trying to get this to work, but it's a bit of a challenge because of all the different colour schemes. However, I think this is possible by using an rgba value.

<img width="278" alt="Screenshot 2020-05-12 at 16 32 07" src="https://user-images.githubusercontent.com/43215253/81713792-38e3fb80-946e-11ea-89c0-470a8a906eb1.png">

That way, you avoid the issue of difficulties to see the background in other schemes.

<img width="254" alt="Screenshot 2020-05-12 at 16 33 31" src="https://user-images.githubusercontent.com/43215253/81713902-603ac880-946e-11ea-8629-ef8df18d4d3d.png">

Would that be okay?

davemart-in commented:
> Would that be okay?

Yep, that sounds like a great solution. Thanks!

Aurorum committed:
Make requested style changes

Aurorum commented:
This PR should be ready for review again! I've checked the different colour schemes and verified there's nothing that looks off, and that the text wraps at least somewhat better should there be an especially large translation. 

<img width="262" alt="Screenshot 2020-05-12 at 18 16 27" src="https://user-images.githubusercontent.com/43215253/81725256-7a2fd780-947d-11ea-9af7-92e5c440f6b6.png">

a8ci18n commented on client/my-sites/sidebar/index.jsx:
```diff
@@ -183,7 +198,16 @@ export class MySitesSidebar extends Component {
+				{ ! hideChecklistProgress && (
+					<div className="sidebar__checklist-progress">
+						<p className="sidebar__checklist-progress-text">
+							{ translate( 'Setup', { context: 'verb' } ) }
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **30** times:
`translate( 'Setup', { context: 'Filter label for theme content'} )` *ES Score: 11*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/pull/42022)


:information_source: This string already exists with the following **contexts**:
- `null` (no context)
- `Filter label for theme content`

 Would it make sense to reuse one of the above?

kwight commented:
Looks fantastic, and great solution for the color schemes 🎉

rickybanister commented:
It could be nice to make the progress bar narrower as @davemart-in suggests above, but I also noticed for any sites with traffic (even your own traffic) you'll see the sparkline just below the progress bar:

![image](https://user-images.githubusercontent.com/437258/81755953-4e661f00-9488-11ea-8d5a-a5c6a428730a.png)

The sparkline is hardcoded to 80px. Since the two bars are in such close proximity it may make sense to align them and use 80px for both.

Aurorum commented:
Sounds good, PR for that in #42143. :)

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/3555940

Thank you @Aurorum for including a screenshot in the description! This is really helpful for our translators.
